docs/RFCS/20200811_non_blocking_txns.md

```diff
@@ -0,0 +1,674 @@
+
+If we start committing transactions in the future, this guarantee is no longer
+trivially satisfied. If we want to ensure linearizability for the writes of
+these non-blocking transactions, we need to do a little extra work to ensure
+that the clocks on all nodes are sufficiently close to the commit timestamp of
+the transactions before acknowledging their success to clients.
+
+Conveniently, we already have a formalism for this – `hlc.Update`. Commit Wait
+would naturally fall out of calling `hlc.Update(txn.CommitTimestamp)` before
+acknowledging any transaction commit to a client. This would reduce to a no-op
+for standard transactions, and would result in a wait of up to
+`non_blocking_duration` for non-blocking transactions. This wait artificially
+increases the latency of non-blocking transactions, but critically, it ensures
+that we continue to preserve linearizability.
+
+It is worth noting that, like Spanner, this CommitTimestamp is chosen before the
+final round of consensus replication. This means that the wait will overlap with
+consensus communication and will therefore be less disruptive to non-blocking
+transactions than it may initially seem.
+
+#### Uncertainty Intervals: To Wait or Not to Wait
+
+Non-blocking transactions also force us to rethink uncertainty intervals and
+question both how and why they work. Uncertainty intervals are timestamp ranges
+within which a transaction is unable to make claims of causality. Given a
+bounded clock offset of `max_offset` between nodes, a transaction cannot
+definitively claim that a write that it observes within a `max_offset` window on
+either side of its original timestamp if causally ordered before it. To ensure
+that all possible causal dependencies are captured and linearizability is
+enforced, a transaction will bump its timestamp to ensure that it observes all
+values in its uncertainty interval. This either incurs a transaction refresh or
+a full transaction restart.
+
+The writes performed by non-blocking transactions are no exception. When a read
+observes a write performed by a non-blocking transaction in its uncertainty
+interval, it will need to bump its timestamp so that it observes the write.
+
+The complication here is that if a reading transaction bumped its timestamp
+above a value written by a non-blocking transaction in its uncertainty interval,
+it could end up with a timestamp in the future of "present time". This could
+risk resulting in a stale read if this read immediately committed and then
+triggered a causally dependent read that hit a slower gateway where the written
+value was not considered in the second read's uncertainty interval. However,
+careful readers may notice that if the first read-only transaction respected the
+Commit Wait rule then it would have been delayed before committing and the stale
+read would not be possible.
+
+But relying on Commit Wait here violates a desirable property of our transaction
```

Comment:
I hope I'm misunderstanding something here, but isn't this basically saying that our read latencies will spike to ~30ms (uncertainty interval) for each non-blocking write when it becomes visible?

For read-only txns (or non-blocking read-write txns) I hoped we would not need to apply uncertainty to writes that it knows came from a non-blocking txn, but it doesn't seem to quite work. If the non-blocking writer committed the intents only after making sure that the commit timestamp is no longer synthetic (i.e. its clock is past that timestamp), then we're resolving intents too late and readers contend. If we resolve them early and then wait, the readers have to deal with uncertainty.

See if you can polish this section a bit - I had some trouble following "second read", "first read-only transaction", maybe an explicit itemized example with timestamps would be helpful.

Reply:
Hmm, I think/hope I'm just confused here. If it's t=100 now and a transaction future-writes-and-commits at t=1000, only readers with `t \in [970,999)` have to worry about uncertainty. But if reader's timestamp was `< 1000` when it was chosen (non-synthetically), it can't causally depend on the previous writer if that writer would have waited for its local clock to hit 1030 (future write ts + uncertainty) to ack the commit, which means our dear reader couldn't have started after the writer returned and gotten a timestamp less than 1000.